src/modules/ekf2/EKF/gps_control.cpp

```diff
@@ -112,30 +117,63 @@ void Ekf::controlGpsFusion()
-						const bool was_gps_signal_lost = isTimedOut(_time_prev_gps_us, 1'000'000);
-
-						/* A reset is not performed when getting GPS back after a significant period of no data
-						 * because the timeout could have been caused by bad GPS.
-						 * The total number of resets allowed per boot cycle is limited.
-						 */
-						if (isYawFailure()
-						    && _control_status.flags.in_air
-						    && !was_gps_signal_lost
-						    && _ekfgsf_yaw_reset_count < _params.EKFGSF_reset_count_limit
-						    && isTimedOut(_ekfgsf_yaw_reset_time, 5'000'000)) {
+					const bool is_vel_fusion_failing = isTimedOut(_aid_src_gnss_vel.time_last_fuse, _params.reset_timeout_max);
+					const bool is_pos_fusion_failing = isTimedOut(_aid_src_gnss_pos.time_last_fuse, _params.reset_timeout_max);
+
+					// A reset is not performed when getting GPS back after a significant period of no data because the timeout could have been caused by bad GPS.
+					// The total number of resets allowed per boot cycle is limited.
+					const bool was_gps_signal_lost = isTimedOut(_time_prev_gps_us, 1'000'000);
+
+					const bool yaw_failure = isYawFailure() && !was_gps_signal_lost
+								 && (_ekfgsf_yaw_reset_count < _params.EKFGSF_reset_count_limit)
+								 && isTimedOut(_ekfgsf_yaw_reset_time, 5'000'000);
+
+					const bool should_reset_gps_fusion = shouldResetGpsFusion();
+
+					if (should_reset_gps_fusion || (yaw_failure && (is_vel_fusion_failing || is_pos_fusion_failing))) {
+
+						bool yaw_reset = false;
+
+						if (yaw_failure) {
 							// The minimum time interval between resets to the EKF-GSF estimate is limited to allow the EKF-GSF time
 							// to improve its estimate if the previous reset was not successful.
 							if (resetYawToEKFGSF()) {
 								ECL_WARN("GPS emergency yaw reset");
+								yaw_reset = true;
+
+								if (_control_status.flags.mag_hdg || _control_status.flags.mag_3D) {
+									// stop using the magnetometer in the main EKF otherwise it's fusion could drag the yaw around
+									// and cause another navigation failure
+									_control_status.flags.mag_fault = true;
+									_warning_events.flags.emergency_yaw_reset_mag_stopped = true;
+								}
+
+								if (_control_status.flags.gps_yaw) {
+									_control_status.flags.gps_yaw_fault = true;
+									_warning_events.flags.emergency_yaw_reset_gps_yaw_stopped = true;
+								}
+
+								if (_control_status.flags.ev_yaw) {
+									// Stop the vision for yaw fusion and do not allow it to start again
+									stopEvYawFusion();
+									_inhibit_ev_yaw_use = true;
+								}
 							}
 						}
 
-						ECL_WARN("GPS fusion timeout, resetting velocity and position");
-						_information_events.flags.reset_vel_to_gps = true;
-						_information_events.flags.reset_pos_to_gps = true;
-						resetVelocityTo(gps_sample.vel, vel_obs_var);
-						resetHorizontalPositionTo(gps_sample.pos, pos_obs_var);
+						if (should_reset_gps_fusion || is_vel_fusion_failing || yaw_reset) {
```

Comment:
Skip velocity reset if `_aid_src_gnss_vel` timeout, but `_time_last_hor_vel_fuse` is still recent?